autorelay: Split libp2p.EnableAutoRelay into 2 functions

sukunrt opened the pull request:
Provide two specific ways to enable the autorelay subsystem libp2p.EnableAutoRelayWithStaticRelays
libp2p.EnableAutoRelayWithPeerSource

fixes: https://github.com/libp2p/go-libp2p/issues/1866

marten-seemann commented on options.go:
```diff
@@ -315,6 +319,31 @@ func EnableAutoRelay(opts ...autorelay.Option) Option {
+}
+
+// EnableAutoRelayWithPeerSource configures libp2p to enable the AutoRelay subsystem using
+// the provided peerSource callback to get more relay candidates
```
```suggestion
// the provided peerSource callback to get more relay candidates.
```

marten-seemann commented on options.go:
```diff
@@ -315,6 +319,31 @@ func EnableAutoRelay(opts ...autorelay.Option) Option {
 }
 
+// EnableAutoRelayWithStaticRelays configures libp2p to enable the AutoRelay subsystem using
+// the provided slice of relays as relay candidates
```
```suggestion
// the provided relays as relay candidates.
```

marten-seemann commented on p2p/host/autorelay/options.go:
```diff
@@ -174,3 +174,12 @@ func WithClock(cl clock.Clock) Option {
+
+// WithMinInterval sets the minimum interval after which peerSource callback will be called for more
+// candidates even if AutoRelay needs new candidates
+func WithMinInterval(interval time.Duration) Option {
```
Why is the minimum interval both a parameter to `WithPeerSource` and a separate option here?

sukunrt replied:
I want to fix that, just wanted to confirm once if it's fine to change WithPeerSource method signature. 

https://github.com/libp2p/go-libp2p/issues/1866#issuecomment-1406184775

sukunrt replied:
fixed. Removed from WithPeerSource

marten-seemann reviewed (commented):
One questions about the API. There's also a failing staticcheck check.

sukunrt commented:
@marten-seemann addressed review comments. 
I've removed the test TestIncorrectInit since that should be fixed by this PR and shouldn't need this test that checks for panic on incorrect initialisation.

marten-seemann reviewed (approved):
LGTM. @MarcoPolo, could you have a look at this PR please? Just want to make sure we definitely get the API right this time.

MarcoPolo commented:
Pushed a change to make the `PeerSource` function a separate type so that it can be documented properly

marten-seemann commented on p2p/host/autorelay/options.go:
```diff
@@ -10,9 +10,22 @@ import (
+// call the callback again until the channel is closed. Implementations should
+// send new peers, but may send peers they sent before. AutoRelay implements a
+// per-peer backoff (see WithBackoff). See WithMinInterval for setting the
+// minimum interval between calls to the callback. The context.Context passed
+// MAY be canceled when AutoRelay feels satisfied, it will be canceled when the
+// node is shutting down. If the channel is canceled you MUST close the output
+// channel at some point.
+type PeerSource func(ctx context.Context, num int) <-chan peer.AddrInfo
```
I don't think this is correct. We're passing in a `context.Context` here, not the `context.CancelFunc`, so AutoRelay has no way of canceling this context.

MarcoPolo replied:
AutoRelay is the one passing in the context. There is a typo in this comment. It should say If the context is canceled instead of if the channel is canceled.

marten-seemann replied:
Right, sorry for the stupid mistake.

marten-seemann commented on p2p/host/autorelay/options.go:
```diff
@@ -10,9 +10,22 @@ import (
+// send new peers, but may send peers they sent before. AutoRelay implements a
+// per-peer backoff (see WithBackoff). See WithMinInterval for setting the
+// minimum interval between calls to the callback. The context.Context passed
+// MAY be canceled when AutoRelay feels satisfied, it will be canceled when the
```
This is not a RFC 2119 "may".
```suggestion
// may be canceled when AutoRelay feels satisfied, it will be canceled when the
```
